### ISSUE TITLE
[Bug]: Composition plotting error

### ISSUE THREAD

**Issue by matthewsmith-NOAA** [r4ss/r4ss#770]
### Describe the bug

SS version 3.30.20
R4ss version 1.47.0 from github

When running SS_plots plotting proceeds as usual until reaching group 13. In group 13, items 21-24 are produced by ssplotComps and output after which the code crashes and returns "Error in if (CompError == 1) { : argument is of length zero". 

R4ss functions were run on a converged model a few days ago and worked as expected (unknown r4SS version). R4ss was updated yesterday and now produces the above error on the same model output. 

[3.Hybrid.zip](https://github.com/r4ss/r4ss/files/10561624/3.Hybrid.zip)

### To Reproduce

Unzip and extract files. Use ss_output to input report file into R without covar (file was too large to attach) or run the model with ss version 3.30.20 to create covar file. Use SS_plots to generate plots. observe error when script reaches plot group 13.

### Expected behavior

I expected the full suite of plots to be produced, exported to the directory on my PC, and the HTML files to be created.

### Which OS are you seeing the problem on?

Windows

### Which versions of r4ss and SS3 are you seeing the problem on?

r4SS 1.47.0, SS3.30.20

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/r4ss/r4ss#code-of-conduct)

**Comment by iantaylor-NOAA** [comment]
Thanks @matthewsmith-NOAA for reporting this. I introduced this bug earlier this week via an inadequately tested changes related to the in-development version of SS3 where separate Dirichlet-multinomial parameters will be available for discards vs retained fish, where that SS3 change required a different table format.

Your model works for me after the change I just made in commit 6ee9723, so please update r4ss (sometimes also requires restarting R), and let me know if it doesn't work for you. If the fix works, go ahead and close this issue.

**Comment by iantaylor-NOAA** [comment]
Update: I accidentally pushed the fix to the branch I was working on, not "main". That branch is now merged via Pull Request #771, so the fix wasn't available via `remotes::install_github("r4ss/r4ss")` 5 minutes ago when I last posted here, but should now be available.

**Comment by matthewsmith-NOAA** [comment]
Thanks Ian! updated package this morning and plots work as expected. Appreciate the quick turn around on the issue.

**Comment by lisaailloud-NOAA** [comment]
Hi @iantaylor-NOAA - I downloaded the latest version of r4ss from github (r4ss version 1.48.0) and I am getting the exact same plotting issue as Matt (SS version V3.30.21.00) when using SS_plots() function on a model that has Dirichlet parameters.

_Starting length comp data plots (group 13)
Error in if (CompError == 1) { : argument is of length zero_

**Comment by iantaylor-NOAA** [comment]
Hi @lisaailloud-NOAA, sorry about this.
I was able to replicate the error using the "3.Hybrid.zip" model attached to the original post on this issue, but I haven't had time to figure out a solution yet. I will work on it ASAP.

**Comment by nathanvaughan-NOAA** [comment]
Hey Ian, if it helps I did some digging around for Lisa on this and the crash is happening on line 1478 of the SSplotComps.R file and appears to be because in her case data_info[["partition"]] == j and data_info[["Fleet"]] == f have no overlapping entries for some combinations so sub ends up being an empty data frame and CompError becomes an empty length 0 (lines 1468:1475). I think the easiest work around would be to add a if(length(CompError)!=0){} wrapper around lines 1477:1521 with an else option that puts in a default caption when the CompError length is zero. I just wasn't sure if there were other follow on issues this would cause. Hope this helps speed things up for you.

**Comment by iantaylor-NOAA** [comment]
Thanks @nathanvaughan-NOAA. That's definitely the problem and the solution would work as a temporary fix, but may miss the correct DM parameter. The table of comp specifications including CompError in the original model has rows and values for all fleets, and gets returned by SS_output() as $Age_comp_error_controls, but somehow by the time it gets to that section of SSplotComps(), some of the rows are missing and I don't know why. I won't have time to work on this until late afternoon EDT.

**Comment by iantaylor-NOAA** [comment]
@lisaailloud-NOAA and @nathanvaughan-NOAA, I think this is now sorted out in the `comp_plotting_error_770` branch of r4ss which you can install via `pak::pkg_install("r4ss/r4ss@comp_plotting_error_770")`.

I'm on leave the rest of the week but will check in about this and the checks on PR #801 on Monday.

**Comment by nathanvaughan-NOAA** [comment]
Thanks @iantaylor-NOAA we will test it out. Hope you enjoy your leave :)

**Comment by lisaailloud-NOAA** [comment]
@iantaylor-NOAA - I tested your fix and I am still getting an error but it's partly my fault! I did a few test runs and narrowed down the issue:

I only get the error if Dirichlet parameters are being used & I've entered partition==2 when in fact it should've been zero. 


If I enter partition==2 in the data file when really it should be a zero, SS will change it automatically and return a warning: 

> Warning 6 : fleet: 1  lencomp has obs with partition==2; will treat as partition=0 because retention not defined; N= 35

I've not paid much attention to this warning in the past because SS "fixes the problem" for me and it hadn't caused any other issues, but in this specific case where Dirichlet parameters are being used, it will cause an error with r4ss. I don't know if that's something that should be addressed in R4SS or if it would be better addressed on the SS side? (I don't know if others, like me, tend to ignore that error message?)

To summarize (regardless of whether I use `install_github("r4ss/r4ss")` or the `comp_plotting_error_770` branch) : 
- if I have Dirichlet parameters & wrongly enter partition==2 instead of 0 -- I get the error
- if I **do not** have Dirichlet parameters & wrongly enter partition==2 instead of 0 -- I **do not** get the error
- if I have Dirichlet parameters & I correctly enter partition==0 --  I **do not** get the error

The fix is not pressing since I can just correct the partition input on my end to get it to run without error, so no rush there! I hope this helps - let me know if you need more info from me.

**Comment by nathanvaughan-NOAA** [comment]
Thanks @lisaailloud-NOAA  for the detailed explanation, that matches up with the cause of the error that I found digging through the code. I wasn't sure if the partition=2 in the data was an error or not but that is definitely the cause. Make sense now that it doesn't fail for the no Dirichlet scenario if SS automatically fixes it to be 0. I wonder if inputting the 2 is causing any errors in the results?

**Comment by lisaailloud-NOAA** [comment]
@nathanvaughan-NOAA - that's right, SS treats it as a zero internally. I double checked just to be sure and my two runs (with part input 0 and 2) yield identical results.

**Comment by iantaylor-NOAA** [comment]
Hi @lisaailloud-NOAA, good to hear you narrowed down the problem and found a work-around.

For completeness it would be good to get this bug sorted out eventually. The files I was using (attached to original post on this issue) have a different setup on discards than what you're working with now, so could you email me the files that cause the r4ss error ("Dirichlet parameters & wrongly enter partition==2")?

**Comment by nathanvaughan-NOAA** [comment]
@iantaylor-NOAA I haven't had a chance to look at the cause yet (and probably won't for a while) but I also got an error when running SS_readclt for this assessment 

Error in if (age_selex_Nparms[j] > 0) { : 
  missing value where TRUE/FALSE needed

I assume they have a similar source so this is probably an issue that is useful to solve inside SS (given that it is actually caused by a data file error) at least to give a more pressing warning output that this setup is likely to cause errors in r4ss even though SS can make it work internally. I would say SS could overwrite the partitions in data.ss_new with zeros but that would probably create more issues than it is worth.

**Comment by iantaylor-NOAA** [comment]
I think the fixes to SS3 have resolved this issue old issue with SS3 version 3.30.21 so I closed the old pull requests #799 and #801 without merging any changes to r4ss. If folks have any further issues related to this feel free to reopen.
